allow alternate Parabola instantiation

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* geometry
  * Parabola can be instantiated from an expression in terms of x and/or y
<!-- END RELEASE NOTES -->

smichr committed:
allow alternate Parabola instantiation

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -30,6 +30,11 @@ class Parabola(GeometrySet):
         Default value is Point(0, 0)
     directrix : Line
+    apq : tuple of ``a`` and x intercepts
+    ahk : tuple of ``aa`` and vertex ``(h, k)``
```
```suggestion
    ahk : tuple containing ``a``, ``h``, ``k`` for ``a*(x - h)**2 + k``
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -30,6 +30,11 @@ class Parabola(GeometrySet):
     focus : Point
         Default value is Point(0, 0)
     directrix : Line
+    apq : tuple of ``a`` and x intercepts
```
```suggestion
    apq : a tuple containing ``a``, ``p``, ``q`` for ``a*(x - p)*(x - q)``
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -30,6 +30,11 @@ class Parabola(GeometrySet):
     directrix : Line
+    apq : tuple of ``a`` and x intercepts
+    ahk : tuple of ``aa`` and vertex ``(h, k)``
+    abc : tuple of coefficients for ``a*x**2 + b*x + c``
```
```suggestion
    abc : tuple containing ``a``, ``b``, ``c`` for ``a*x**2 + b*x + c``
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -60,14 +65,86 @@ class Parabola(GeometrySet):
+    >>> from sympy.abc import x, y
+    >>> p1.intersection(Line(x))  # y axis where x = 0
+    [Point2D(0, 4)]
+    >>> p1.intersection(Line(y))  # y axis where y = 0
```
```suggestion
    >>> p1.intersection(Line(y))  # x axis where y = 0
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -60,14 +65,86 @@ class Parabola(GeometrySet):
+
+    >>> from sympy import solve
+
+    >>> Parabola(eq=x**2 - 4)
+    Parabola(Point2D(0, 0), Line2D(Point2D(4, 0), Point2D(4, 1)))
```
```suggestion
    >>> Parabola(eq=x**2/4)
    Parabola(Point2D(0, 1), Line2D(Point2D(0, -1), Point2D(1, -1)))
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -60,14 +65,86 @@ class Parabola(GeometrySet):
+    >>> solve(_, y)
+    [x**2 - 5*x + 6]
+    >>> _[0].factor()
+    (x - 2)*(x - 3)
```
```suggestion
    (x - 3)*(x - 2)
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -60,14 +65,86 @@ class Parabola(GeometrySet):
+            if sum(1 for i in hints if i) != 1:
+                raise ValueError('pass a single other instantiation keyword')
+            if eq is not None:
+                eq = eq.rewrite(Add)
```
```suggestion
                eq = eq.lhs - eq.rhs
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -60,14 +65,86 @@ class Parabola(GeometrySet):
+    >>> p1.intersection(Line(y))  # x axis where y = 0
+    [Point2D(-8, 0), Point2D(8, 0)]
+
+    Alternate instantiation from a tuple of parameters is also possible:
+
+    >>> from sympy import solve
+
+    >>> Parabola(eq=x**2/4)
+    Parabola(Point2D(0, 1), Line2D(Point2D(0, -1), Point2D(1, -1)))
```
```suggestion
   The Parabola may be described by a quadratic expression:

    >>> Parabola(eq=x**2/4)  # or y - x**2/4, Eq(y, x**2/4)
    Parabola(Point2D(0, 1), Line2D(Point2D(0, -1), Point2D(1, -1)))
    >>> Parabola(eq=y**2 + 2*y + 3)
    Parabola(Point2D(9/4, 1), Line2D(Point2D(7/4, 0), Point2D(7/4, 1)))

    A tuple of parameters can also be used to create the Parabola:

    >>> from sympy import solve
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -60,14 +65,86 @@ class Parabola(GeometrySet):
+
+    >>> Parabola(abc=(1, 2, 3)).equation(x, y).expand()
+    -x**2 - 2*x + y - 3
+    >>> solve(_, y)[0]
```
```suggestion
    >>> solve(_, y)[0]  # see that y(x) = (1)*x**2 + (2)*x + (3)
```

smichr commented:
@sylee957 , @dimitra-kon , do you have any suggestions? Anyone else that might use this in an educational setting?

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -5,9 +5,9 @@
 """
 
-from sympy.core import S
+from sympy.core import S, Add
```
```suggestion
from sympy.core import S
```

sylee957 commented:
I may suggest to split the initialization into 4 functions, and have `parabola_from_roots, parabola_from_quadratic_standard_form`,
than populating the constructor.

Anyway I don't think that `apq, abc, ...` are good names to use. Maybe I'd like to consult the ChatGPT about the standard terminology.

smichr committed:
allow focus to act as eq

This is similar to how Line acts:

>>> Line(2*x + y - 3) == Line((0,3),(1,1)) == Line((0,3),slope=-2)
Line2D(Point2D(0, 3), Point2D(1, 1))

smichr commented:
I would like to keep a coherent behavior for geometry: see last commit note.

Since conventions can vary while equations cannot, it now accepts an equation, not a tuple of parameters. This avoids ambiguity like `4*p*(x-h)**2 + k` vs `a*(x-h)+k`.

smichr commented:
> I may suggest to split the initialization into 4 functions

In the future, I would like to have a GeometricEntity class that accepts up to a second order expression and returns the appropriate object, e.g GeometricEntity((x-1)**2 + (y-2)**2)->Point(1, 2)`

smichr committed:
Update parabola.py

smichr commented:
@sylee957 , how strongly do you feel about not adding this?

sylee957 commented:
I don't think that we should reuse the `focus` parameter like `eq_or_focus = sympify(focus)`. 
This can give semantically wrong usage like `Parabola(focus=x**2 + 2*x + 3)`

I think that alternative from parsing the equations, we can ask users to inputs standard form like

`Parabola.from_standard_form(a, b, c, d, e, f)`

and this is better because `as_coefficients_dict` may gives the ambiguous solution than user wants.
For instance, `((x - 1)*(x - 2) + 3).as_coefficients_dict()` gives `{1: 3, (x - 2)*(x - 1): 1}`, 
so we endlessly get an argument whether or not to expand the equation

smichr commented:
Thanks for your comments. I was hoping that entering an expression was unambiguous, however, since you can't instantiate a Point from a non-iterable. But it is possible that a user might accidentally enter `x**2` for the focus instead of `(x**2, 0)` and then get a Parabola instead of an error. So I agree.

What about a keyword `coeff` instead of a different method?

oscarbenjamin commented:
> What about a keyword `coeff` instead of a different method?

I prefer having different functions or methods rather than keyword or optional arguments. It is generally better if each individual function has a simple, strict, unambiguous signature.

Functions with complicated signatures are hard to change or improve as well as hard to understand. Any time a signature is made more complex we make it harder to improve anything in future as well as being confusing for users.

smichr committed:
allow abc instantiation

smichr commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -141,3 +141,15 @@ def test_parabola_intersection():
+
+def test_Parabola_alt_instantiation():
+    ans = -x**2 - 3*x/2 + y/2 - S(5)/2
+    eq = 2*x**2 + 3*x + 5
+    assert Parabola(eq).equation(x, y).expand() == ans
+    assert Parabola(eq.subs(x, y)).equation(y, x).expand() == ans  # reversed x and y
```
```suggestion
    eq = 2*x**2 + 3*x + 5
    assert Parabola(abc=(2, 3, 5)).equation(x, y).expand() == ans
    assert Parabola(abc=(2, 3, 5)).rotate(-pi/2).equation(y, x).expand() == ans  # reversed x and y
```

smichr commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -1,6 +1,6 @@
 from sympy.core.numbers import (Rational, oo)
```
```suggestion
from sympy.core.numbers import (pi, Rational, oo)
```

smichr commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -141,3 +141,15 @@ def test_parabola_intersection():
     raises(TypeError, lambda: parabola1.intersection(2))
+
+
+def test_Parabola_alt_instantiation():
```
```suggestion
def test_is_similar():
    assert not Parabola().is_similar(Line((0, 0), slope=1)
    assert Parabola().is_similar(abc=(1,2,3))


def test_Parabola_alt_instantiation():
```

smichr committed:
Apply suggestions from code review

smichr commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -141,3 +141,20 @@ def test_parabola_intersection():
+
+
+def test_is_similar():
+    assert not Parabola().is_similar(Line((0, 0), slope=1)
```
```suggestion
    assert not Parabola().is_similar(Line((0, 0), slope=1))
```

smichr committed:
Update sympy/geometry/tests/test_parabola.py

smichr commented:
> Functions with complicated signatures are hard to change or improve as well as hard to understand

Please check the current version. If this is still frowned upon then I will suggest `Parabola.abc(1, 2, 3)`.

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -5,9 +5,9 @@
 """
 
-from sympy.core import S
+from sympy.core import S, sympify
```
```suggestion
from sympy.core import S
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -47,32 +49,77 @@ class Parabola(GeometrySet):
 
-    >>> from sympy import Parabola, Point, Line
-    >>> p1 = Parabola(Point(0, 0), Line(Point(5, 8), Point(7,8)))
+    >>> from sympy import Parabola, Point, Line, Eq
```
```suggestion
    >>> from sympy import Parabola, Point, Line
```

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -420,3 +467,7 @@ def vertex(self):
+
+    def is_similar(self, other):
+        # all parbolas are self similar
+        return is_instance(other, Parabola)
```
```suggestion
        return isinstance(other, Parabola)
```

smichr committed:
Apply suggestions from code review

sylee957 commented on sympy/geometry/parabola.py:
```diff
@@ -47,32 +49,77 @@ class Parabola(GeometrySet):
+
+    def __new__(cls, focus=None, directrix=None, abc=None, **kwargs):
+
+        if directrix is not None:
+            if abc is not None:
+                raise ValueError('enter focus & directrix or abc')
+            focus = Point(focus or (0, 0), dim=2)
+            directrix = Line(directrix)
+            if directrix.contains(focus):
+                raise ValueError('The focus must not be a point of directrix')
+        else:
+            if focus is not None:
+                raise ValueError('enter focus & directrix or abc')
```
I think that the specific reason `a signature is made more complex` is that revealed here.

For example, you are branching first with `directrix` and then 
in the nested if statement, you are writing the code that assumes something eliminate the logically negative cases (`directix is None`)

So it is literally the case
```
if focus is not None and directrix is not None and abc is None:
    ...
if focus is None and directrix is None and abc is not None:
    ...
```

Of course, it may not be that much difficult to understand this at the point now,
but later modifying/understanding the code should be equivalent to solving some satisfiability problem about the logics here, and that's one thing I'd like to comment about here.

smichr replied:
That's a good point and I wonder if you raised that some years ago. Every decision in a routine adds to the complexity both of testing and "digesting". That's good enough reason to go for a dedicated `abc` method.
